File: test/render_recipe_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:rana/features/render/model/render_recipe.dart';

void main() {
  test('legacy v0 recipe migrates to v1 with bounded editable values', () {
    final recipe = RenderRecipeV1.fromMap(const <String, dynamic>{
      'tone': 140,
      'textureVal': -5,
      'undertoneX': 4,
      'presetId': 'portra',
    });

    expect(recipe.tone, 100);
    expect(recipe.texture, 0);
    expect(recipe.undertoneX, 1);
    expect(recipe.presetId, 'portra');
    expect(recipe.toMap()['recipeVersion'], currentRenderRecipeVersion);
  });

  test('recipe round trips without losing wire parameters', () {
    const recipe = RenderRecipeV1(
      temperature: 0.2,
      colorMatrix: <double>[1, 0.1, 0, 0, 1, 0, 0, 0.2, 1],
      lightLeakVariant: 2,
      aspectRatio: 'square_1_1',
      outputQuality: 'heic',
      presetId: 'test',
    );

    expect(RenderRecipeV1.fromMap(recipe.toMap()), recipe);
  });

  test('unknown version produces a structured non-destructive error', () {
    expect(
      () =>
          RenderRecipeV1.fromMap(const <String, dynamic>{'recipeVersion': 99}),
      throwsA(
        isA<UnsupportedRenderRecipeVersion>()
            .having((error) => error.code, 'code', 'UNSUPPORTED_RECIPE_VERSION')
            .having((error) => error.version, 'version', 99),
      ),
    );
  });
}

File: android/app/src/test/kotlin/com/rana/app/rana/RenderRecipeTest.kt
package com.rana.app.rana

import org.junit.Assert.assertEquals
import org.junit.Assert.assertThrows
import org.junit.Test

class RenderRecipeTest {
    @Test
    fun `legacy flat recipe migrates to v1 and clamps invalid style values`() {
        val recipe = RenderRecipeV1.fromMap(
            mapOf(
                "tone" to 140.0,
                "textureVal" to -2.0,
                "undertoneX" to 3.0,
                "presetId" to "portra"
            )
        )

        assertEquals(100f, recipe.tone)
        assertEquals(0f, recipe.texture)
        assertEquals(1f, recipe.undertoneX)
        assertEquals("portra", recipe.presetId)
        assertEquals(RENDER_RECIPE_VERSION, recipe.toMap()["recipeVersion"])
    }

    @Test
    fun `v1 recipe round trips through its compatibility map`() {
        val original = RenderRecipeV1(
            temperature = 0.2f,
            colorMatrix = listOf(1f, 0.1f, 0f, 0f, 1f, 0f, 0f, 0.2f, 1f),
            lightLeakVariant = 2,
            aspectRatio = "square_1_1",
            outputQuality = "heic",
            presetId = "test"
        )

        assertEquals(original, RenderRecipeV1.fromMap(original.toMap()))
    }

    @Test
    fun `unknown recipe versions fail without discarding metadata`() {
        val error = assertThrows(UnsupportedRenderRecipeVersionException::class.java) {
            RenderRecipeV1.fromMap(mapOf("recipeVersion" to 99))
        }

        assertEquals(99, error.recipeVersion)
    }
}
